src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java

```diff
@@ -2,70 +2,104 @@
- *   determined by the command line parser via a call to isDependentArgumentAllowed).
- *
- * NOTE: plugin class @Arguments that are marked "optional=false" should be not have a primitive
- * type, and should not have an initial value, as the command line parser will interpret these as
- * having been set even if they have not been specified on the command line. Conversely, @Arguments
- * that are optional=true should have an initial value, since they parser will not require them
- * to be set in the command line.
- *
- * The methods for each descriptor are called in the following order:
- *
- *  getPluginClass()/getPackageNames() - once when argument parsing begins (if the descriptor
- *  has been passed to the command line parser as a target descriptor)
- *
- *  getClassFilter() - once for each plugin subclass found
- *  getInstance() - once for each plugin subclass that isn't filtered out by getClassFilter
- *  validateDependentArgumentAllowed  - once for each plugin argument value that has been
- *  specified on the command line for a plugin that is controlled by this descriptor
- *
- *  validateArguments() - once when argument parsing is complete
- *  getAllInstances() - whenever the pluggable class consumer wants the resulting plugin instances
- *
- *  getAllowedValuesForDescriptorArgument is only called when the command line parser is constructing
- *  a help/usage message.
+ * <p>
+ * Base class for a descriptor that defines a set of command line plugin classes that can be dynamically
+ * discovered by the command line parser. Plugin instances contain argument fields that extend the
+ * set of command line arguments accepted by the parser.
+ * </p>
+ * <p>
+ * Each plugin descriptor passed to the command line parser is queried to determine the class and
+ * package names to be searched for plugin classes controlled by that descriptor. The descriptor
+ * object itself is  added to the parser's list of argument sources, and for each matching plugin class
+ * discovered in the packages list defined the descriptor, the parser delegates back to the descriptor
+ * to obtain an instance of that plugin. Each of these plugin objects are then also added to the parsers's
+ * list of argument sources.
+ * </p>
+ * <p>
+ * Descriptor instances should have at least one @Argument field used to accumulate the plugins that
+ * are specified on the command line by the user. Allowed values for this argument are taken from the
+ * plugins that have been dynamically discovered.
+ * </p>
+ * <p>
+ * A descriptor implementation may optionally choose to provide a constructor that specifies a list of
+ * default plugin instances that are automatically included along with any plugins that are specified
+ * manually by the user. The descriptor and command line parser generally treat these default instances
+ * as if they had been specified by the user as arguments on the command line. </li>
```

Comment:
The last sentence in this paragraph should be a caveat to solve in the future. Maybe it should be noted that it will change because it is a bug described in https://github.com/broadinstitute/barclay/issues/23

Reply:
I'm not sure what we would say about this here. If and when we fix the bug, I think the only change here would be that this last sentence would be removed.

Reply:
But I think that it is useful to say that it is a bug that will be solved, and not a feature done in purpose. Like that, API users will be aware that it may change in the future...